exp/topology/desiredstate/desired_state.go

```diff
@@ -526,6 +527,11 @@ func (g *generator) computeControlPlaneVersion(ctx context.Context, s *scope.Sco
 	}
 
 	if feature.Gates.Enabled(feature.RuntimeSDK) {
+		if annotations.HasWithPrefix(clusterv1.BeforeClusterUpgradeHookAnnotationPrefix, s.Current.Cluster.Annotations) {
+			log.Info(fmt.Sprintf("Cluster upgrade to version %q is blocked by %q annotation hook", desiredVersion, clusterv1.BeforeClusterUpgradeHookAnnotationPrefix))
```

Comment:
Let's make this more precise. Please see what we have in the Machine controller

```go
			var hooks []string
			for key := range m.ObjectMeta.Annotations {
				if strings.HasPrefix(key, clusterv1.PreDrainDeleteHookAnnotationPrefix) {
					hooks = append(hooks, key)
				}
			}
			slices.Sort(hooks)
			log.Info("Waiting for pre-drain hooks to succeed", "hooks", strings.Join(hooks, ","))
```

(your current message is fine and consistent with below, but I would like to include the exact hooks that are blocking and get rid of the prefix constant)

Reply:
I would suggest
```go
			var hooks []string
			for key := range s.Current.Cluster.Annotations {
				if strings.HasPrefix(key, clusterv1.BeforeClusterUpgradeHookAnnotationPrefix) {
					hooks = append(hooks, key)
				}
			}
			slices.Sort(hooks)
			log.Info(fmt.Sprintf("Cluster upgrade to version %q is blocked by %q hook (via annotations)", desiredVersion, runtimecatalog.HookName(runtimehooksv1.BeforeClusterUpgrade)), "hooks", strings.Join(hooks, ","))
			return *currentVersion, nil
```
(I think conceptually this is also a BeforeClusterUpgrade hook)